### ISSUE TITLE
Nix is completely broken after major macOS update

### ISSUE THREAD

**Issue by kballard** [NixOS/nix#2444]
After updating macOS to 10.14 Mojave, Nix is completely broken, giving me an error saying something about I think libSystem_networking.dylib being missing (I apologize for not having the exact error text, I already blew away my local installation).

My recollection is I went through the same issue when upgrading from macOS 10.12 to 10.13, where I had to completely reinstall Nix and my Nix-installed packages.

It would be fantastic if this issue did not happen again next year when upgrading to macOS 10.15.

**Comment by edolstra** [comment]
It's pretty much a certainty that it *will* break on 10.15, because, well, that's macOS :-)

**Comment by domenkozar** [comment]
Maybe install script should safeguard macos to be set to known tested versions?

**Comment by kballard** [comment]
That would be awful. The *only* remedy to a macOS upgrade breaking Nix is to reinstall Nix. Locking the installer script to known versions would prevent that, without a good reason, as reinstalling works just fine.

**Comment by kahlil29** [comment]
Having the same issue.
Reinstalled but still getting the following error (that I got after upgrading on my older installation)

```
dyld: Library not loaded: /usr/lib/system/libsystem_network.dylib
  Referenced from: /nix/store/wrl2sxy6yqgm7w7ii1515jxndncc7545-Libsystem-osx-10.11.6/lib/libSystem.B.dylib
  Reason: image not found
```

Completely stuck here.

**Comment by domenkozar** [comment]
@kahlil29 did you upgrade to Nix 2.1?

**Comment by kahlil29** [comment]
@domenkozar I nuked my installation and ran it again from scratch via
```
curl https://nixos.org/nix/install | sh
```

It installed Nix version 2.1.2

**Comment by kahlil29** [comment]
Is there no known workaround for this? 
I tried the solution given here as well with the tar file by @LnL7  [Temp Solution](https://github.com/NixOS/nix/issues/2244#issuecomment-399660394) but that ended up giving me the same error.

I have now resorted to running an Ubuntu VM on my system and had to install everything from scratch.

**Comment by LnL7** [comment]
The issue you describe was fixed in 2.1. I have not upgraded myself, but I have helped somebody with the upgrade process so whatever this is it's something specific to your system. Either something that isn't 2.1 or something from the previous install is interfering.

**Comment by kahlil29** [comment]
@LnL7 Thanks for your input on this!
I'll try nuking everything properly once more and re-installing Nix 2.1 again.

**Comment by ajbouh** [comment]
I just had this same issue. Doing the following at least allowed installation to complete without any `Library not loaded` errors.

```
sudo mv /nix /nix-old
export PATH=/usr/bin:/bin:/usr/sbin:/sbin
unset SSL_CERT_FILE
unset NIX_PATH
# ... Reran installer
```

**Comment by kahlil29** [comment]
It's not my nix installation procedure that's triggering the error. Installation runs fine but the moment I go into my company's GitLab repo and run `nix-shell -A shells.ghc` ...it begins the download process and during that is where it comes up with the error(s). Still not able to resolve it.

**Comment by vai0** [comment]
I am encountering the same exact issue as @kahlil29. Ran through the steps to do a clean install. Installation runs just fine - it's just once I run `nix-shell` within my project it throws:

```
dyld: Library not loaded: /usr/lib/system/libsystem_network.dylib
  Referenced from: /nix/store/46qj01lcjz48wb4hc7mvn3mhb3syda7h-Libsystem-osx-10.11.6/lib/libSystem.B.dylib
  Reason: image not found
```

**Comment by LnL7** [comment]
What you describe sounds a lot more like an expression that uses a pinned nixpkgs that didn't have 10.14 support yet instead of an issue with the install.

**Comment by jordanst3wart** [comment]
This does actually seem to be an issue with the individual packages. Reinstalling the individual packages fixes the issue for me (see below).

```
unknown at unknown in ~
$ micro
dyld: Library not loaded: /usr/lib/system/libsystem_network.dylib
  Referenced from: /nix/store/12yshvbpz9hjn7khzzf0ar67kaw6g7f7-Libsystem-osx-10.11.6/lib/libSystem.B.dylib
  Reason: image not found
[1]    10652 abort      micro

unknown at unknown in ~
$ nix-env -e micro
uninstalling 'micro-1.4.0'
building '/nix/store/scgi90x34jj9f6f2mw6yj532hzhsd5jk-user-environment.drv'...
created 465 symlinks in user environment

unknown at unknown in ~
$ nix-env -i micro
warning: name collision in input Nix expressions, skipping '/Users/jordanstewart/.nix-defexpr/channels_root/nixpkgs'
replacing old 'micro-1.4.0'
installing 'micro-1.4.1'
these paths will be fetched (5.76 MiB download, 14.25 MiB unpacked):
  /nix/store/4w5hz6sh4924p2wqn2zp2gsc03wqd83p-micro-1.4.1-bin
  /nix/store/lss6icbzgvl1ns2ns74l56sxqdngwhpq-micro-1.4.1
copying path '/nix/store/4w5hz6sh4924p2wqn2zp2gsc03wqd83p-micro-1.4.1-bin' from 'https://cache.nixos.org'...
copying path '/nix/store/lss6icbzgvl1ns2ns74l56sxqdngwhpq-micro-1.4.1' from 'https://cache.nixos.org'...
building '/nix/store/3a822vr0xn0n6isc2ka4619j5w8xiln2-user-environment.drv'...
created 469 symlinks in user environment

unknown at unknown in ~
$ micro

unknown at unknown in ~
$ nix-env --version
nix-env (Nix) 2.1.3
```

**Comment by s50600822** [comment]
pretty sure going for 2.1.3 will get https://github.com/NixOS/nix/issues/2193 instead.

**Comment by jbddc** [comment]
> I am encountering the same exact issue as @kahlil29. Ran through the steps to do a clean install. Installation runs just fine - it's just once I run `nix-shell` within my project it throws:
> 
> ```
> dyld: Library not loaded: /usr/lib/system/libsystem_network.dylib
>   Referenced from: /nix/store/46qj01lcjz48wb4hc7mvn3mhb3syda7h-Libsystem-osx-10.11.6/lib/libSystem.B.dylib
>   Reason: image not found
> ```

I'm also getting this error and already nuked my nix install and installed it again. :/

**Comment by s50600822** [comment]
Hi if https://github.com/NixOS/nix/issues/2518 is addressed, I can use latest Nix on macOS can anyone help ?

**Comment by jbddc** [comment]
> @jordanst3wart yeah, if you use `nix-env`, packages will stay at whatever version they were at when you installed them until you upgrade them, even if you update your channel. If that version was built on a foundation that was incompatible with Mojave, they'll break.
> 
> @jbddc @kahlil29 we need more information to debug something like that. What expression led to that? Do you have any idea which program is crashing with that error? There might be crash logs in Console.app for example.

Well, I can confirm that I can install packages using `nix-env`, which I suppose calls the most recent channel of `nixpkgs-unstable`.
What I was trying to install, however, was [haskellR](https://tweag.github.io/HaskellR/docs/build-and-install.html) using the `--nix` flag with stack. I subsequently got the aforementioned `library not loaded` error. 

If the issue is related to the main project (haskellR that is) having an older version of nixpkgs pinned, then I think this workflow is not sustainable if nix is to be adopted by the masses. macOS upgrades breaking packages like this makes me think that either:

1. The de facto way of writing nix expressions for darwin packages is wrong
2. The way that nix abstracts macOS system libraries and files could be improved to take into account its dynamic shift of location from upgrade to upgrade
3. Apple just likes to mess with developers and completely change it's OS's "infrastructure" every year :/

**Comment by drever** [comment]
I ran into the same problem when trying to install [miso](https://github.com/dmjio/miso#nix). I seems to be pinned to `Libsystem-osx-10.11.6`. I there any way to fix this?

**Comment by kahlil29** [comment]
@copumpkin I think my issue was due to `bytestring-trie` 
In a more recent version there has been some change to do with `monoid` and `semigroup` that causes the version I have not to build with GHC `8.4.3` 
I think the fix is to use a patched version of the package locally, but I haven't got down to trying it yet, hopefully in the next few days. 
Check this issue : https://github.com/reflex-frp/reflex-platform/issues/398
So in conclusion, my issue is nothing nix-specific, from what I gather.

**Comment by newhoggy** [comment]
I'm experience this issue:

```
$ git clone git@github.com:mankyKitty/fantastic-waddle.git
$ cd fantastic-waddle
$ nix build -f release.nix
builder for '/nix/store/1mj7i40s6h363y10b2w9cgfjr5jr64ff-stdenv-darwin-boot-0.drv' failed due to signal 6 (Abort trap: 6); last 3 log lines:
  dyld: Library not loaded: /usr/lib/system/libsystem_network.dylib
    Referenced from: /nix/store/cp0dhigj0lxg96d5m5qmdws4k417jvy2-bootstrap-tools/lib/libSystem.B.dylib
    Reason: image not found
```

**Comment by kitlangton** [comment]
I'm also experiencing this issue

**Comment by mtesseract** [comment]
@kitlangton For me a similar build failure was fixed by updating the project's pinned Nixpkgs version.
This is what @vai0 also hinted at.

**Comment by kitlangton** [comment]
I'm still experiencing this issue, nothing seems to fix it. Has anyone whose experience this issue been able to fix it?

**Comment by chreekat** [comment]
I think this issue is fixed in Nixpkgs, but can someone point to the sha where it got fixed?

**Comment by koalalorenzo** [comment]
I just upgraded to a minor version and got a similar issue.

**Comment by grahamc** [comment]
Hey y'all, I don't usually talk about the Determinate Nix Installer here, since I think it is typically off topic. However, in this case I think it might help a lot of people. We've recently [fixed this problem in the Determinate Nix Installer](https://determinate.systems/posts/nix-survival-mode-on-macos), which you can get like this:

```sh
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix | sh -s -- install
```

Note it does have some minor differences from the upstream Nix installer, which you can read about in its README: https://github.com/determinateSystems/nix-installer#installation-differences.

**Comment by motorlearner** [comment]
Possibly related: 

I am on macOS 15.6.1 and I am using a `flake.nix` with hardcoded `nixpkgs` revision to make a devshell with R. This worked just fine 2 weeks ago. But now, after running `nix develop` and then `R --version`:

```
dyld[4663]: Library not loaded: /nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib/libgfortran.5.dylib
  Referenced from: <41397F32-E2BD-3244-A926-A788C5CDF38E> /nix/store/8ixwcraqmw518n84ix5spzsfmac05jd2-R-4.4.1/lib/R/lib/libR.dylib
  Reason: tried: 
'/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib/libgfortran.5.dylib' (duplicate LC_RPATH '/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib'),
'/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib/libgfortran.5.dylib' (duplicate LC_RPATH '/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib'), 
'/System/Volumes/Preboot/Cryptexes/OS/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib/libgfortran.5.dylib' (no such file), 
'/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib/libgfortran.5.dylib' (duplicate LC_RPATH '/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib'), 
'/nix/store/8ixwcraqmw518n84ix5spzsfmac05jd2-R-4.4.1/lib/R/lib/libgfortran.5.dylib' (no such file), '/nix/store/k8h7h0y3m19iaydkfhjz32mlhr07rlgp-zulu-ca-jdk-21.0.2/zulu-21.jdk/Contents/Home/lib/server/libgfortran.5.dylib' (no such file)
zsh: abort      R
```

I have no idea what is going on, it appears MacOS refuses to load `libgfortran.5.dylib`. Is this a SIP issue?

It works fine on some other refs. The ref I use, where it fails, is `12228ff1752d7b7624a54e9c1af4b222b3c1073b`.

**Comment by lilyball** [comment]
@motorlearner What were you running 2 weeks ago? I'm assuming you updated macOS since you're commenting on this issue. From searching around it appears that as of macOS 15.4, having duplicate `LC_RPATH`s is considered an error and will cause libraries to fail to load.

If I look at `/nix/store/d3da9lm0fglqw6l8d9124md7rwdqlcwm-gfortran-13.3.0-lib/lib/libgfortran.5.dylib`, it has a bunch of duplicated LC_RPATH entries. It has 2 unique paths that are duplicated 5 times each. So that's why it's being rejected now. If I look at the version of that library in my currently-installed nixpkgs, it only has a single LC_RPATH entry (which is just `@loader_path`). So the problem has been fixed since the version of nixpkgs you're using.

**Comment by motorlearner** [comment]
@lilyball Aaah thank you, that must be it. There was indeed an automatic update, before I was on 15.0.1.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
